### ISSUE TITLE
Trilinos: Transition Trilinos to require c++14 standard

### ISSUE THREAD

**Issue by rppawlo** [trilinos/Trilinos#6260]
Transition Trilinos to require c++14 standard rather than c++11. Many apps are already building with c++14 and Kokkos would benefit greatly from this. This ticket is to outline requirements and start discussions with stakeholders. 

Compiler support:
Aggregate table of compiler support: https://en.cppreference.com/w/cpp/compiler_support

Minimum compiler version that has full support for c++14:
- gcc: 5.0 (https://gcc.gnu.org/projects/cxx-status.html#cxx14)
- clang: 3.4 (https://clang.llvm.org/cxx_status.html)
- intel: 17.0 
   - supports everything except N3664: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2013/n3664.html
- ibm xl c/c++: 16.1 
- Visual Studio: VS 2017, 15.7 (https://docs.microsoft.com/en-us/cpp/overview/visual-cpp-language-conformance?view=vs-2019)

From a SEMS standpoint, it looks like we are in good shape.

To do:
* [ ] Discussion with SART teams. Any blockers?
* [ ] Check on SEMS module support for all compilers
* [ ] Set up test builds 
* [ ] Switch PR testing to c++14 
* [ ] ?

@jwillenbring @bartlettroscoe @jmgate @crtrott @mwglass @maherou @prwolfe @trilinos/framework

**Comment by mhoemmen** [comment]
The last check box is definitely "profit" -- it would be pleasant to have auto lambdas and not to have to reimplement `make_unique`.

**Comment by mwglass** [comment]
Here’s a short list of where a correct C++11 program can break under a C++14 compliant compiler:

- Digit Separators

The digit separator was introduced so that one could, in a more readable manner, write numeric literals and split them up in a more natural way.

- Sized Deallocation

C++14 introduces the opportunity to declare a global overload of operator delete suitable for sized deallocation, something which wasn't possible in C++11.

- constexpr member-functions, no longer implicitly const

There are many changes to constexpr in C++14, but the only change that will change semantics between C++11, and C++14 is the constantness of a member-function marked as constexpr.

- Removal of std::gets

std::gets has been removed from the Standard Library because it is considered dangerous.

**Comment by swbova** [comment]
from the ALEGRA team's perspective, C++14 compliance, and the associated compiler upgrades has not at all been on our radar and was not part of our planning process for FY20.    It might not be a big deal, or it could be a huge deal across our entire TPL stack--we just don't know.  I would like to see this deferred until FY21.

**Comment by mhoemmen** [comment]
@swbova It's quite likely that nothing will need to change other than the compiler flags you use in the build system.  C++11 to C++14 is a very low-risk move.  I would budget more time for the move to C++17, as that changes semantics in more significant ways (e.g., copy elision).

**Comment by mwglass** [comment]
A good first step would be to make sure Trilinos is C++14 compliant to meet the needs of code teams that are currently moving to the use of C++14. The second step would be to develop a timeline for when Trilinos would require C++14. I understand hat for Kokkos this is still about a year away. Is Trilinos going to support both C++11 or C++14 during some transition period? That would certainly take care of the the ALEGRA team's concerns.

**Comment by swbova** [comment]
@mhoemmen, I understand the code changes may be minimal.  At the same time, ALEGRA has 4,000 nonlinear multiphysics tests. if just 10% of them diff, say due to order of operations issues, across one or two (platform,compiler) then it becomes a significant effort to determine that the code is still producing acceptable answers.

@mwglass, yes, this would address my concerns.

**Comment by crtrott** [comment]
@mwglass regarding transition: if your definition of transition is "support both" we are in transition for C++20. Kokkos supports everything from 11 through 20 already, and has been for a bit.

**Comment by mhoemmen** [comment]
@swbova As Mike and Christian said, I think it would make sense first to test Kokkos and Trilinos with both C++11 and C++14.  That would ensure that Trilinos could work with both.  At that point, ALEGRA could decide whether it wanted to try testing with C++14.

**Comment by rppawlo** [comment]
Action Items:
1. Jason to check with xyce team: coe, sems, ...
2. Jim to check with sdk team - currently requires gcc 4.8
3. Christian to follow up with Alegra team

**Comment by rppawlo** [comment]
> Action Items:
> 
>     1. Jason to check with xyce team: coe, sems, ...
> 
>     2. Jim to check with sdk team - currently requires gcc 4.8
> 
>     3. Christian to follow up with Alegra team

@jmgate @jwillenbring and @crtrott - any updates on the items above?

**Comment by jmgate** [comment]
I would guess the Jason intended there is Verley.

**Comment by rppawlo** [comment]
Sorry about that @jmgate . Just sent an email to Jason V.

**Comment by kddevin** [comment]
While Trilinos remains at c++11, perhaps @trilinos/framework could set up a Trilinos build with c++14, so that we could remove any problems that would block SIERRA's use of c++14.  @jwillenbring @ZUUL42  would a c++14 build be feasible?

**Comment by mhoemmen** [comment]
@kddevin If we do that, I think it's a good idea to have two release builds (with all the optimizations turned on) that differ only by C++ language version.  If applications like ALEGRA have diffs, it would be easier to debug them if that were the only difference.

**Comment by jwillenbring** [comment]
CC: @ZUUL42 @rppawlo 

Consistent with @kddevin 's comment above, at the Product Leader meeting today we discussed turning on the C++14 standard for the GCC 8.3 build, while still supporting back to C++11 (the GCC 4.8.4 build will do this as long as it is around).

I want to verify here that the right change to make is to add:

``
-DCMAKE_CXX_FLAGS="-std=c++14"
``

Are there any other Trilinos-specific options needed? It sounded like that was it for now, but I want to confirm so this is done right.

**Comment by bartlettroscoe** [comment]
> I want to verify here that the right change to make is to add:
>
> `-DCMAKE_CXX_FLAGS="-std=c++14"`

FYI: After the merge of PR #6814, this will become:

```
-DCMAKE_CXX_STANDARD=14
```

I am in the process of testing that now against the ATDM APP codes. (Unfortunately, that breaks strict backward compatibility.)

**Comment by jwillenbring** [comment]
The PR to move the GCC 8.3 build from C++11 to C++14 has been merged. I also noticed that since GCC 6, C++14 has been the default, so based on that, all this time the GCC 7.2 build has been running with C++14.

To be clear, Trilinos still supports the the C++11 standard at this time, this effort was associated only with making sure things also work with C++14 so our customers can build with C++14. This does not allow unguarded usage of C++14 in Trilinos.

**Comment by bartlettroscoe** [comment]
> I also noticed that since GCC 6, C++14 has been the default, so based on that, all this time the GCC 7.2 build has been running with C++14.

@jwillenbring, I don't think that is true.  What is the evidence for you saying that?  The current TriBITS system explicitly sets `-std=c++11` so all builds of Trilinos were/are using C++11 explicitly by default, not C++14. You can see this, for example, in the build `PR-6976-test-Trilinos_pullrequest_gcc_4.9.3_SERIAL-422` shown [here]() which shows the compiler arguments with `-std=c++11`.  (You can't see that on CDash for the gcc_7.2.0 build because there are no build warnings for that build and therefore you can't see any compile lines.)

NOTE: Once we merge PR #6814 (which I am in the process of testing now), all Trilinos builds will be, by default, setting `CMAKE_CXX_STANDARD=11` so, again, all Trilinos builds will be using `-std=c++11` explicitly by default.  (But the user will be able to set `-DCMAKE_CXX_STANDARD=14` or `17` or `20`, etc. and therefore this GCC 8.3 PR build will need to be setting `-DCMAKE_CXX_STANDARD=14` instead.)

**Comment by jwillenbring** [comment]
@bartlettroscoe 

OK, I did not realize TriBITS was doing this. I was just citing the GCC documentation for the default. I tried to check the compile line for the build I was looking at yesterday too, but there were no warnings so I could not.

So are you saying now then that the GCC 8.3 build, despite specifying -std=c++14 in CMAKE_CXX_FLAGS is actually not enabling C++14 due to TriBITS?

**Comment by bartlettroscoe** [comment]
>  I tried to check the compile line for the build I was looking at yesterday too, but there were no warnings so I could not.

@jwillenbring, just a tip, but if you look at the generated {{build.ninja}} file, you can clearly see the compile options for each target. 

> So are you saying now then that the GCC 8.3 build, despite specifying -std=c++14 in CMAKE_CXX_FLAGS is actually not enabling C++14 due to TriBITS?

Hum, if you look at the compile line on CDash (which does have a warning) shown [here](https://testing.sandia.gov/cdash/viewBuildError.php?type=1&buildid=6534995), you will see the compile line:

`"/projects/sems/install/rhel6-x86_64/sems/compiler/gcc/8.3.0/openmpi/1.10.1/bin/mpicxx" "-DUSING_CMAKE" "-I." "-I/scratch/trilinos/workspace/trilinos-folder/Trilinos_pullrequest_gcc_8.3.0/Trilinos/packages/seacas/libraries/exodus/include" "-Ipackages/seacas/libraries/exodus" "-isystem" "/projects/sems/install/rhel6-x86_64/sems/tpl/netcdf/4.4.1/gcc/8.3.0/openmpi/1.10.1/exo_parallel/include" "-isystem" "/projects/sems/install/rhel6-x86_64/sems/tpl/hdf5/1.8.12/gcc/8.3.0/openmpi/1.10.1/parallel/include" "-pedantic" "-Wall" "-Wno-long-long" "-Wwrite-strings" "-Wshadow" "-Woverloaded-virtual" "-std=c++14" "-fno-strict-aliasing" "-Wall" "-Wno-clobbered" "-Wno-vla" "-Wno-pragmas" "-Wno-unknown-pragmas" "-Wno-parentheses" "-Wno-unused-local-typedefs" "-Wno-literal-suffix" "-Wno-deprecated-declarations" "-Wno-misleading-indentation" "-Wno-int-in-bool-context" "-Wno-maybe-uninitialized" "-Wno-class-memaccess" "-Wno-inline" "-Wno-nonnull-compare" "-Wno-address" "-DTRILINOS_HIDE_DEPRECATED_HEADER_WARNINGS" "-std=c++11" "-O3" "-DNDEBUG" "-MD" "-MT" "packages/seacas/libraries/exodus/test/CMakeFiles/SEACASExodus_ExoIICTests.dir/ExoIICTests.cxx.o" "-MF" "packages/seacas/libraries/exodus/test/CMakeFiles/SEACASExodus_ExoIICTests.dir/ExoIICTests.cxx.o.d" "-o" "packages/seacas/libraries/exodus/test/CMakeFiles/SEACASExodus_ExoIICTests.dir/ExoIICTests.cxx.o" "-c" "packages/seacas/libraries/exodus/test/ExoIICTests.cxx"`

and you see the order of the compiler flags:

```
... "-std=c++14"  ... "-std=c++11" ... 
```

That makes me think this is actually a C++11, not a C++14 build.

No no one actually look at the compile line before they set up that PR build?

**Comment by bartlettroscoe** [comment]
> OK, I did not realize TriBITS was doing this

This is documented at:

* https://tribits.org/doc/TribitsBuildReference.html#enabling-support-for-c-11

And before people criticize TriBITS for this, remember, `<Project>_ENABLE_CXX11` and `<Project>_CXX11_FLAGS` was added years before CMake added support for `CMAKE_CXX_STANDARD` (or we were allowed to use it since we were stuck on CMake 2.8.11 until just over a year ago).  For example, there is no `CMAKE_CXX_STANDRD` in CMake 3.0 as (not) shown in:

* https://cmake.org/cmake/help/v3.0/search.html?q=CMAKE_CXX_STANDARD&check_keywords=yes&area=default

**Comment by TBird2001** [comment]
I'm commenting on behalf of the Xyce team. We appear to have a path forward for supporting C++14, so we will not be a blocker for this issue.

**Comment by ibaned** [comment]
Just to relay discussion from the ALEGRA team, they would like to see some testing of Trilinos with C++14 before trying it, but are otherwise open to it.

**Comment by rppawlo** [comment]
Here's the current status of c++14 in Trilinos:

After the next release (3.2), Kokkos will be requiring C++14.  Related to this, the Kokkos 3.3 release and the subsequent integration with Trilinos will require a C++14 compliant compiler.  This is a heads up that the tentative date for this C++14 requirement in Trilinos is September 28, 2020.  After that, Trilinos developer and master will require C++14.  Previous releases of Trilinos (13 and before) should continue to work with C++11.

**Comment by jwillenbring** [comment]
Only the CUDA build has a few remaining issues. We will also have to turn off the GCC 4.8.4 build when we move to C++14. Note that per @crtrott the estimated date for Kokkos requiring C++14 in Trilinos has been changed to early November:

The snapshot we are doing is a bug fix release for the last release. It does NOT require C++14. C++14 requirement will come with Kokkos 3.3 likely early November.

**Comment by etphipp** [comment]
Is there a specific date yet as to when Trilinos will require C++14?

**Comment by rppawlo** [comment]
@etphipp - we set the date as the end of September 2020 to coincide with kokkos promotion into trilinos that requires it. The kokkos promotion got delayed so they didn't throw the switch yet. We could probably do that now if needed. @jwillenbring ?

**Comment by etphipp** [comment]
> @etphipp - we set the date as the end of September 2020 to coincide with kokkos promotion into trilinos that requires it. The kokkos promotion got delayed so they didn't throw the switch yet. We could probably do that now if needed. @jwillenbring ?

I'm putting some stuff in Sacado right now that could benefit from C++14, but there are easy workarounds for C++11.  I was just wondering how soon I would be allowed to use C++14 in an unguarded manner.

**Comment by jwillenbring** [comment]
@etphipp 
@rppawlo is right saying that the Kokkos promotion was delayed. We had/have all of the barriers removed in the PR testing. So the technical constraints should be gone. It is more of a policy issue now. Since the Kokkos change should be coming soon and 2 product leads have asked also, I feel it is reasonable to go ahead and push your changes, realizing that if you are the first, we may need to keep an eye on customer builds and any straggling issues with PR testing for anything that wasn't caught during planning.

**Comment by etphipp** [comment]
OK.  It looks to me that Trilinos is still using C++11 though.  Grepping through the Trilinos cmake code I see that C++14 is specified in all of the PR testing scripts by setting CMAKE_CXX_STANDARD=14, but according to Tribits C++11 is still the default:
```
  # Set up C++ language standard selection.
  IF (NOT CMAKE_CXX_STANDARD)
    SET(CMAKE_CXX_STANDARD 11)
  ELSEIF (NOT CMAKE_CXX_STANDARD MATCHES "^(11|14|17|20)$")
    MESSAGE(FATAL_ERROR "CMAKE_CXX_STANDARD=${CMAKE_CXX_STANDARD} is not 11, 14, 17, or 20.")
  ENDIF ()
```
Don't we need to change the default to C++14 and actually not allow C++11 as a choice?  Otherwise we would still have to guard our use of C++14.

**Comment by kddevin** [comment]
@jwillenbring I propose we throw the switch and change the default.  We advertised well the Sept 28 target date, and the PR testing is all good.  It is OK if we make this change ahead of Kokkos.

**Comment by bartlettroscoe** [comment]
@etphipp and @rppawlo, can we schedule a short meeting offline about how to handle C++ standards in Trilinos?  See my notes below.

----

The variable ` ${PROJECT_NAME}_ENABLE_CXX11` is no longer supported.  Instead, you set CMAKE_CXX_STANDARD.  See:

*  https://github.com/trilinos/Trilinos/blob/4b7012f3e967b5c2b8cb1b1a8ab1f2282d4ed12f/cmake/tribits/ReleaseNotes.txt#L36 

If the user sets `-DCMAKE_CXX_STANDARD=14`, then this will definitely work.  See:

* https://github.com/trilinos/Trilinos/blob/4b7012f3e967b5c2b8cb1b1a8ab1f2282d4ed12f/cmake/tribits/core/package_arch/TribitsGlobalMacros.cmake#L2048

But looking at that code, I am not sure of the best way for Trilinos to specify its C++ standard.

Brad King wrote this TriBITS code but it looks like he did not provide any developer or user documentation for it.  And the behavior as implemented does not look to have the flexibility you want to specify things.  I think there is some missing documentation on how to manipulate C++ standards with TriBITS and the logic should probably be updated to make this more flexible and natural.  There are some tricky nuances about how to specify C++ standards with CMake that I should probably discuss with some knowledgeable Trilinos developers before we pull the trigger on this.

**Comment by etphipp** [comment]
@bartlettroscoe @rppawlo Sure.  Today is no good.  My calendar is up to date.

**Comment by rppawlo** [comment]
My calendar is up to date too.

**Comment by jwillenbring** [comment]
> @jwillenbring I propose we throw the switch and change the default. We advertised well the Sept 28 target date, and the PR testing is all good. It is OK if we make this change ahead of Kokkos.

@kddevin This seems reasonable. I went to make the change to the default and ran into a question about an old C++11 option that I asked @bartlettroscoe about. It is not clear if it should be left, changed or removed. We are working this out now.

**Comment by bartlettroscoe** [comment]
If we just want to get C++14 set right now by default, we can just set that in the base `CMakeLists.txt` file before the `TRIBITS_PROJECT()` call.   I can put in a PR for that if you want.  But we may tweak this once I talk to a few people about handling C++ standards with CMake and Trilinos.

**Comment by bradking** [comment]
From external discussion with @bartlettroscoe:
> target_compile_features(mylib PUBLIC cxx_std_17)
>
> then all downstream C++ source files will get compiled with -std=c++17 automatically.
> But what if in a downstream CMakeLists.txt file they set:
>
> set(CMAKE_CXX_STANDARD 14)

The higher of the two standards will be used.  The `CXX_STANDARD` value
will be superseded by the `cxx_std_17` usage requirement, and the code
that links to `mylib` will compile as C++17.

Basically the `CMAKE_CXX_STANDARD` value says to use the given standard
in place of the compiler's default standard level.  However, if that level is not
sufficient to satisfy the `cxx_std_*` usage requirements, it may be raised for those.

> This is a real use case that came up recently where we are trying to
> build all of Trilinos (including Kokkos) with C++17 but want to build
> the STK package with C++14 (because it will not build with C++17).  But
> STK depends on Kokkos which would have
>   `target_compile_features(mylib PUBLIC cxx_std_17)`
> set on its libraries and therefore propagate to its header files and into
> STK source file builds.

If Kokkos headers use C++17 code then all consumers need `cxx_std_17` to
be able to compile.  If Kokkos headers do not use C++17, or have conditions
to be able to compile as C++14 too, then `cxx_std_17` should not be a public
usage requirement of Kokkos.

**Comment by bartlettroscoe** [comment]
> The higher of the two standards will be used.

@bradking, happens what to downstream customers when `INTERFACE_COMPILE_FEATURES cxx_std_14` is set on all libraries in Trilinos but someone sets `-D CMAKE_CXX_STANDARD=17` when building Trilinos (which would mean that `-std=c++17` would get passed to the compiler)?  What value for `INTERFACE_COMPILE_FEATURES` will get put into the generated `<Package>Targets.cmake` file (which gets included in the `<Package>Config.cmake` file)?  Will it be `std_cxx_14` or `cxx_std_17`? 

The issue is that there are ifdefs in Trilinos header files that depend on the C++ standard being used by the compiler as determined by the standard compiler macro `__cplusplus`.  So the header files my behave differently when compiled with `-std=c++14` vs. `-std=c++17`.    For example, there is logic like:

```
./seacas/packages/seacas/libraries/suplib_cpp/fmt/core.h:383:#if __cplusplus >= 201703L  // C++17's char_traits::length() is constexpr.

./packages/stk/stk_performance_tests/stk_mesh/GearsSkinning.cpp:215:#if __cplusplus >= 201402L

./packages/rol/src/zoo/ROL_Types.hpp:921:#if __cplusplus >= 201402L // using C++14
```

And there is preprocessor logic all over GTest like this:

```
googletest/include/gtest/internal/gtest-port.h:2268:#if __has_include(<any>) && __cplusplus >= 201703L
```

One could avoiding this problem by configuring and installing `<Package>_config.h` files that remember the C++ standard used in the upstream CMake build and removing all preprocessor logic based on the value of `__cplusplus` in the header files.  But that would require editing several C++files in Trilinos and putting in configure-time logic in configured and installed `<Package>_config.h` files.

Any thoughts/advice on this?

**Comment by bradking** [comment]
> `INTERFACE_COMPILE_FEATURES cxx_std_14` is set on all libraries in Trilinos but someone sets `-D CMAKE_CXX_STANDARD=17` when building Trilinos...What value for INTERFACE_COMPILE_FEATURES will get put into the generated `<Package>Targets.cmake` file 

The generation of `INTERFACE_COMPILE_FEATURES` is fully determined by `target_compile_features` calls.  `CMAKE_CXX_STANDARD` never affects what is propagated as outgoing usage requirements.  In your example, the `<Package>Targets.cmake`  will get `cxx_std_14` and not `cxx_std_17`.

CMake doesn't know what ABI changes projects make in response to different C++ standard levels.  It is possible to provide a C++14 ABI from sources using C++17 in their implementation.  CMake only provides methods for projects to publish this information themselves.  Trilinos could be taught to check `CMAKE_CXX_STANDARD`'s value and add extra public `cxx_std_##` feature requirements to match it.

> One could avoiding this problem by configuring and installing <Package>_config.h files that remember the C++ standard used in the upstream CMake build and removing all preprocessor logic based on the value of __cplusplus in the header files.

Alternatively, add `INTERFACE_COMPILE_DEFINITIONS` that tell your headers what features are enabled in their libraries.

**Comment by ibaned** [comment]
@bradking how dow `target_compile_features` interact with extensions? Doing just `target_compile_features(mylibrary PUBLIC cxx_std_17)` actually puts `-std=gnu++17` onto the command line which I'd rather not have since I don't want to accidentally start using GCC extensions, I'd like to get `-std=c++17`. Is the best way to do this still to just `set_property(TARGET mylibrary PROPERTY CXX_EXTENSIONS OFF)` ?

**Comment by bartlettroscoe** [comment]
>The generation of `INTERFACE_COMPILE_FEATURES` is fully determined by `target_compile_features()` calls. `CMAKE_CXX_STANDARD` never affects what is propagated as outgoing usage requirements. 

@bradking, given that is the case, then perhaps the behavior of the code at:

* https://github.com/trilinos/Trilinos/blob/4b7012f3e967b5c2b8cb1b1a8ab1f2282d4ed12f/cmake/tribits/core/package_arch/TribitsGlobalMacros.cmake#L2048

that sets `INTERFACE_COMPILE_FEATURES` to the same C++ standard as `CMAKE_CXX_STANDARD` is the correct thing to do (since the higher standard will always be used anyway so why bother allowing them to be set differently at the global level)?

But if the downstream customer sets `CMAKE_CXX_STANDARD=17` in their CMake project but Trilinos exported libraries that set `INTERFACE_COMPILE_FEATURES` to `cxx_std_14`, then the install Trilinos header files getting pulled into downstream customer code builds will be submitted to `-std=c++17`, correct?  If so, that is bad.  (But again, if we know this, I think we can address that at configure-time upstream with configured and installed `<Package>_config.h` files.)

BTW, are any of these nuances documented anywhere in standard CMake documentation?  It looks like some of this is documented in the book "Professional CMake: A Practical Guide (5th edition)" but some of the nuances being discussed in this GitHub issue don't seem to be covered explicitly in that book either (at least not that I can find).

**Comment by bradking** [comment]
> how does `target_compile_features` interact with extensions?

It is orthogonal.  All the compile features care about is whether the compiler runs in a mode that supports the feature/standard-level.  `cxx_std_14` is satisfied by either `-std=gnu++14` or `-std=c++14`.  The `CMAKE_CXX_EXTENSIONS`/`CXX_EXTENSIONS` is the way to control extensions.

> sets `INTERFACE_COMPILE_FEATURES` to the same C++ standard as `CMAKE_CXX_STANDARD` is the correct thing to do

Yes, and IIRC that is what the `SET(${PROJECT_NAME}_CXX_STANDARD_FEATURE cxx_std_${CMAKE_CXX_STANDARD})` line does.

> if the downstream customer sets `CMAKE_CXX_STANDARD=17` in their CMake project but Trilinos exported libraries that set `INTERFACE_COMPILE_FEATURES` to `cxx_std_14`, then the install Trilinos header files getting pulled into downstream customer code builds will be submitted to `-std=c++17`, correct?

Yes.

IMO it's poor practice to change the project's ABI in header files based on the C++ standard level.  Those ABI changes should be activated by explicit preprocessor definitions, chosen at the time the libraries are built, and propagated to consumers via `INTERFACE_COMPILE_DEFINITIONS` or `<Package>_config.h` files.

> are any of these nuances documented anywhere in standard CMake documentation?

I don't think so.  The individual pieces are each covered by their corresponding reference documentation, but we don't have any guide-level text calling out these interactions.  I opened [CMake Issue 21395](https://gitlab.kitware.com/cmake/cmake/-/issues/21395) to discuss at least adding notes about these interactions to the reference docs.

**Comment by alanw0** [comment]
@bradking I may have missed some details in the above exchange, but I noticed you said stk won't build with c++17. We would be eager to address whatever specific errors you are seeing. We have a c++17 line on the sierra dashboard and it looks ok. (Although it looks like it failed last night during the trilinos build so I guess I'm not 100% sure it got as far as stk...)

**Comment by bradking** [comment]
>  I noticed you said stk won't build with c++17.

@alanw0 that was in my post as part of quoted text from @bartlettroscoe.  I'm not familiar with the details.  Either way it would of course be nice to fix STK for C++17.

**Comment by bartlettroscoe** [comment]
> IMO it's poor practice to change the project's ABI in header files based on the C++ standard level.

@bradking, and yet that is exactly what GTest does in their header files in many places.  I guess that means that you should never install gtest as a built library for downstream customers to use (like Trilinos is currently doing and SPARC relies on that).

@bradking, thanks for your clarifications on this.

**Comment by bartlettroscoe** [comment]
>  I may have missed some details in the above exchange, but I noticed you said stk won't build with c++17. We would be eager to address whatever specific errors you are seeing.

@alanw0, we should chat off line.

**Comment by bartlettroscoe** [comment]
@bradking, it just occurred to me that from looking at:

* https://cmake.org/cmake/help/v3.13/command/target_compile_features.html#command:target_compile_features
* https://cmake.org/cmake/help/v3.13/prop_tgt/CXX_STANDARD.html?highlight=cxx_standard

it seems that a package could use C++17, for example, in internal code but have its header files may only require C++11 when downstream customers use the headers and libraries.  You would specify that with:

```
set(CMAKE_CXX_STANDARD 17)
add_library(mylib ...)
target_compile_features(mylib PUBLIC cxx_std_11)
```

Is that correct?
 
If that is true, it might be a good idea for TriBITS/Trilinos to allow specifying different C++ standards in `CMAKE_CXX_STANDARD` and `INTERFACE_COMPILE_FEATURES`.  So some Trilinos packages in the near future might require, for example, that they be built with C++17 but downstream customers may only need C++14 when using the header files and linking to the libraries.  Is that a valid use case with the design of these CMake features?

**Comment by bradking** [comment]
> use C++17 in internal code but have its header files may only require C++11 ... You would specify that with: ...

Yes.

> Is that a valid use case with the design of these CMake features?

Yes.

One could even use

```cmake
target_compile_features(mylib PRIVATE cxx_std_17 INTERFACE cxx_std_11)
```

on specific libraries whose sources require C++17 but whose headers can be consumed as C++11.  That would allow `CMAKE_CXX_STANDARD` to be set to a lower value (or not at all) for the project in general, while still meeting `mylib`'s requirements.

**Comment by bartlettroscoe** [comment]
Questions we need to discuss and answer (which I will do with @etphipp and @rppawlo later today):

1. For development and experimentation purposes, do we want to be able to allow `CMAKE_CXX_STANDARD` to be set higher than the C++ standard set by the `INTERFACE_COMPILE_FEATURE` value of `cxx_std_<yr>`?  For example, one could experiment with building Tpetra with C++17 enabled but have downstream clients only have C++14 enabled.  (That seems like a bad thing to support but would it be useful to experiment with scenarios like this?)

2. Do we want to be able to override what  `CMAKE_CXX_STANDARD` and `INTERFACE_COMPILE_FEATURE` `cxx_std_<yr>` are on a package-by-package basis?  For example, would it be useful to try building downstream Trilinos packages like Sacado, Panzer, or experimental packages with a higher C++ standard upstream packages like Kokkos or Tpetra?  (With the current CMake support for C++ standards, you can have downstream code build with a higher C++ standard but not a lower C++ standard.)  This seems like this could be a real use case where more experimental downstream packages could jump to higher C++ standards before more heavily used upstream packages can fully build with higher C++ standards.  (But that would require that header files in upstream Trilinos packages result in identical binary code when subjected to a higher C++ standard compiler flag.)

3. From now on forever are we going to assume that an entire stack of C++ software must be built with the same C++ standard from the lowest level library to the highest level application?  If so, what does that do for software deployment models?

But given that so much of the impacted C++ code in Trilinos is in templated header files (where explicit template instantiation is not being used or just does not make sense), these types of scenarios just may not be feasible or even desirable.  And some of these questions like ` #3` may be out of scope for just considering Trilinos.

Also, it seems like trying to mix C++ standards may just not be portable in general:

* https://stackoverflow.com/questions/38056849/mixing-different-c-standards-with-gcc

So one has to question how portable the CMake approach is of allowing downstream C++ code to build with a higher C++ standard involving header files from upstream code what are matched with libraries built with a lower C++ standard.

(P.S. I posted a question on this to the Spack Slack channel to see if anyone there might know something.)

**Comment by crtrott** [comment]
I think the amount of work to make this actually work if it is at all feasible is not worth it. As you said Ross: there is so much stuff in headers that it might not work at all. We definitely have code biforcation based on the C++ standard on our side in Kokkos. So anything which includes Kokkos header files most likely needs to be compiled with the same standard you set for Kokkos.

**Comment by bartlettroscoe** [comment]
> So anything which includes Kokkos header files most likely needs to be compiled with the same standard you set for Kokkos.

@ibaned, but it seems that CMake will not enforce that and it will allow customers to build and link against Kokkos and the rest of Trilinos with a higher C++ standard without even a warning.

Would it be a good idea to record the C++ standard (i.e. the `__cplusplus` macro value) used to build and install Kokkos in the configured and installed `KokkosCore_config.h` file and then error out at build time if the `__cplusplus` macro does not match that recorded value when any Kokkos header file is involved in a downstream compilation?  Otherwise, we are asking for trouble for our downstream customers.

**Comment by crtrott** [comment]
Yeah we probably should do that.

**Comment by ibaned** [comment]
@bartlettroscoe @crtrott although I'm fine with risk-averse safety checks like that in the headers, I do think we should adopt this `target_compile_features` approach in addition. It is better than what we do now because at least it ensures that everyone that uses (for example) Kokkos will be required to use a standard greater than or equal to what Kokkos used. Right now even that is not enforced. It gets us closer to the goal of enforcing the standard to be equal. It also makes beginner use of Kokkos increasingly easier. And keeping up with idiomatic CMake is important to a healthy relationship with Kitware where they will also be responsive to our needs.

**Comment by bartlettroscoe** [comment]
>  I do think we should adopt this target_compile_features() approach in addition.

@ibaned, but then Kokkos and the rest of Trilinos will need to be careful not to have ifdefs that change the code based on the C++ standard being used (i.e. based on [`__cplusplus`](https://en.cppreference.com/w/cpp/preprocessor/replace#Predefined_macros)).  (Or, decisions about what code to generate must be recorded at configure time in a configured and installed `<Package>_config.h` header file.  That way, the code that is generated will not change depending on what C++ standard is being used by clients that include header files when they compile.)

So either one of two things needs to be done:

**option-1:** Require that Kokkos, Trilinos, and all client code build with the same C++ compiler and the same C++ language standard.  => That should be enforced as described [above](https://github.com/trilinos/Trilinos/issues/6260#issuecomment-722535655).

or

**option-2:** Allow downstream customers to compile with a higher C++ standard => Must record decisions about what code to generate based on the C++ version in a configured `<Package>_config.h` header file (and that header file should also check that `__cplusplus` is >= to what is recorded in the `<Package>_config.h` file).

So which is it, option-1 or option-2?

**Comment by bartlettroscoe** [comment]
>  I do think we should adopt this `target_compile_features()` approach in addition

@ibaned, I should make it clear that TriBITS and Trilinos is already using this approach as seen in:

* https://github.com/trilinos/Trilinos/blob/4b7012f3e967b5c2b8cb1b1a8ab1f2282d4ed12f/cmake/tribits/core/package_arch/TribitsGlobalMacros.cmake#L2053

and here:

* https://github.com/trilinos/Trilinos/blob/4b7012f3e967b5c2b8cb1b1a8ab1f2282d4ed12f/cmake/tribits/core/package_arch/TribitsLibraryMacros.cmake#L867


(This fact is not documented very well but we can fix that.)

So we are not debating whether to use that feature of CMake or not  (we are).  What we are debating is what we want to allow/support for mixing C++ standards in various scenarios described above.

Also note that if your customer is not using CMake, then the `target_compile_features()` approach will not work as a way to enforce a greater or equal C++ standard.  But do we care about customers that don't use  CMake at this point?  (That kind of goes to the discussion of support for export makefiles in https://github.com/TriBITSPub/TriBITS/issues/299 as well which is still  unclear due to issues with Xyce.)

**Comment by ibaned** [comment]
1. Instead of using `#if __cplusplus >= ?` checks to use new features depending on the C++ standard available, packages should only use features available in their minimum supported C++ version. Then this risk can be eliminated entirely. If a package wants to try C++14-only features, it needs to raise its minimum required C++ version to C++14 instead of keeping `#ifdef`'ed code to support C++11.

2. We can still enforce that all users use the same (and not a greater) C++ standard out of an abundance of caution. This is not mutually exclusive with reducing risky behavior that would break in a mixed-standard environment, as in (1).

3. We should not spend energy supporting build systems other than CMake. Maintenance burdens like multiple build system support or multiple standard support through `#if` are threats to our ability to succeed. I know we have production codes that don't use CMake but that should be fixed by having them adopt CMake, not by forcing all dependencies to maintain all build systems.

**Comment by bartlettroscoe** [comment]
> If a package wants to try C++14-only features, it needs to raise its minimum required C++ version to C++14 instead of keeping `#ifdef`'ed code to support C++11.

@ibaned, yes, of course.  No one was suggesting anything different.  And if your clients are using CMake, then CMake will enforce this through `INTERFACE_COMPILE_FEATURES` values `cxx_std_<yr>` attached to the imported library targets read from the `<Package>Config.cmake` file.  It is just that non-CMake customers will not have that enforced.

> We can still enforce that all users use the same (and not a greater) C++ standard out of an abundance of caution.

I think we have to do that given @crtrott's comment [above](https://github.com/trilinos/Trilinos/issues/6260#issuecomment-722531631) and they way that some header files in Trilinos create different code based on the C++ standard being used as reported in `__cplusplus` (see examples [above](https://github.com/trilinos/Trilinos/issues/6260#issuecomment-721884882)).

> We should not spend energy supporting build systems other than CMake.

Fine by me.   Is that the official Kokkos policy?  Can we make that the official Trilinos policy?

**Comment by rppawlo** [comment]
> Is that the official Kokkos policy? Can we make that the official Trilinos policy?

Don't think we can make that decision for Trilinos without talking to internal stakeholders. I think we can get to that point rather quickly though. I do know of 2 codes using export makefiles. I believe both are working on a transition to cmake. Will bring this up at next product leaders meeting.

**Comment by mhoemmen** [comment]
Just as an outsider comment: I had some experience a few months ago with mixing C++ dialects (CLR, a Microsoft extension to C++17) and standards (C++17 and a subset of C++20) in the same application.  It takes discipline not to expose "too new" includes in header files.  I do not recommend this unless you define a "data structure firewall" between the two languages.

**Comment by bartlettroscoe** [comment]
FYI: I talked with @etphipp and @rppawlo and the next step is that I will put in a PR that sets `CMAKE_CXX_STANDARD` to `14` and some checking code to make sure a valid standard is selected and errors out if someone tries to set this to `11`.

As a follow up to this long conversation thread, someone should update `KokkosCore_config.h` (or whatever that file is called) to ensure that the same C++ standard is used when Kokkos header files are included in downstream builds.

**Comment by bartlettroscoe** [comment]
FYI: I posted PR #8317 that changes the default C++ version to C++14 and errors our if you try to set `CMAKE_CXX_VERSION=11` or a lower standard.

**Comment by bartlettroscoe** [comment]
So it will be the policy of Kokkos and Trilinos that the same C++ standard must be used with building Trilinos & Kokkos that is used with all downstream software that uses it?  If so, that should be stated somewhere and it really would be a good idea to enforce that as described [above](https://github.com/trilinos/Trilinos/issues/6260#issuecomment-722535655).

**Comment by bartlettroscoe** [comment]
FYI: Looks like moving to C++14 is breaking some of the ATDM Trilinos builds.  See https://github.com/trilinos/Trilinos/pull/8317#issuecomment-723138268.

Therefore, someone needs to reproduce and fix these build errors before we can turn on C++14 support.

**Comment by ndellingwood** [comment]
@bartlettroscoe 

> FYI: Looks like moving to C++14 is breaking some of the ATDM Trilinos builds. See [#8317 (comment)](https://github.com/trilinos/Trilinos/pull/8317#issuecomment-723138268).

I looked over the kokkos-related errors and commented here https://github.com/trilinos/Trilinos/pull/8317#issuecomment-723147058, those errors occur due to the loaded gcc compiler being too old and not supported when c++14 is enabled (based on the vortex and chama errors). Swapping out the old gcc compiler for gcc version 5.3.0 or newer will resolve the kokkos-related errors (and hopefully knock out most of the downstream package errors).

Something I've noticed is that different compiler version modules on some systems will automatically load different gcc versions that requires some extra vigilance (e.g., I've seen cases where intel/17.0 loads gcc/4.8.4, intel/17.1 loads gcc/4.9.3...), the difference in modules by minor-version caught me off guard so hopefully sharing this is helpful.

**Comment by alanw0** [comment]
FYI, #8323 fixed stk's c++17 build issues, as far as we can tell (by reproducing with gcc 9.3 on sierra lan)

**Comment by prwolfe** [comment]
Vortex is back running and I will be checking on that build today or tomorrow.

**Comment by Xjchen0** [comment]
Hello! I knew currently Trilinos need C++17 standard from [https://trilinos.org/docs/files/TrilinosBuildReference.html](url). But what about the old versions? I think a version mapping table will bring me great convenience as I sometimes use the old versions.

**Comment by bartlettroscoe** [comment]
> Hello! I knew currently Trilinos need C++17 standard from [https://trilinos.org/docs/files/TrilinosBuildReference.html](url). But what about the old versions? I think a version mapping table will bring me great convenience as I sometimes use the old versions.

@Xjchen0, the commit that make the change from C++14 to require C++17 was https://github.com/trilinos/Trilinos/commit/1194d8cd87d45a15b483d96304ddddf17ebd5758 on 1/12/2023.

Comparing that to:

* https://github.com/trilinos/Trilinos/releases

it looks like the most current Trilinos version allowing C++14 was 13.4:

* https://github.com/trilinos/Trilinos/blob/trilinos-release-13-4-branch/CMakeLists.txt

and the next version 14.0 required C++17:

* https://github.com/trilinos/Trilinos/blob/trilinos-release-14-0-branch/CMakeLists.txt

(The documentation did not get updated until some time later.)

**Comment by Xjchen0** [comment]
> > Hello! I knew currently Trilinos need C++17 standard from [https://trilinos.org/docs/files/TrilinosBuildReference.html](url). But what about the old versions? I think a version mapping table will bring me great convenience as I sometimes use the old versions.
> 
> @Xjchen0, the commit that make the change from C++14 to require C++17 was https://github.com/trilinos/Trilinos/commit/1194d8cd87d45a15b483d96304ddddf17ebd5758 on 1/12/2023.
> 
> Comparing that to:
> 
> * https://github.com/trilinos/Trilinos/releases
> 
> it looks like the most current Trilinos version allowing C++14 was 13.4:
> 
> * https://github.com/trilinos/Trilinos/blob/trilinos-release-13-4-branch/CMakeLists.txt
> 
> and the next version 14.0 required C++17:
> 
> * https://github.com/trilinos/Trilinos/blob/trilinos-release-14-0-branch/CMakeLists.txt
> 
> (The documentation did not get updated until some time later.)
> 
> 
> 

Thanks!
